### DISCUSSION TITLE
Idea: archive data / save downloaded Subject Sets for "completed Projects"

### DISCUSSION THREAD

**Discussion by shaunanoordin** [zooniverse/subject-set-search-api#28] [General]
## Efficiency Improvement

Noted as of 8a0594af3d65ec0bfcd2f22ad95e5aff28630c0e
Based on a chat with @snblickhan on Tue 25 Apr 2023 (Sam, I'm tagging you as a minor FYI)

TL;DR / Executive Summary: we have an idea to save on the SSSApi's boot up & rebuild times, by saving/archiving the list of Subjects for "completed Projects" _on this repo_ itself (instead of always fetching the latest list from Panoptes every time)

Observations:
- Booting up the Subject Set Search API can take a very long time. (20-30 mins when running `docker-compose build` locally, similarly when running `lita rebuild subject-set-search-api`)
- Most of this time is taken by the _"download Subjects from Panoptes"_ step (i.e. `python subject-set.py`). There are currently 5 projects to download Subjects from, and there are a _lot_ of Subject Sets.
- If I understand correctly, the _"download Subjects from Panoptes"_ step is required because we need to get a fresh list of Subjects from these Projects.
- 💭 **However, some of these projects are now complete.** For example, Scarlets & Blues (12268) and RGBE Herbarium (12561).
- Datasette data doesn't have to be fetched from the Panoptes source every time; for example, the [Classroom Maps API](https://github.com/zooniverse/classroom-maps-api) pulls its data from CSVs/ZIPs stored on the repo itself.

Idea:
- For the these "completed Projects", do we still need to include those projects in the _"download Subjects from Panoptes"_ step?
- I'm thinking, _no._ If the list of Subjects for these projects are now fixed, we can just store them as CSVs (or CSVs in a ZIP) in this repo.
  - From local tests, Scarlet & Blues and RGBE Herbarium's "download" steps (when run from a laptop in Zootown Oxford) took 8 mins and 10 mins respectively. So archiving these two projects would save quite a bit on the boot up time.
  - Of course, we'd need to find a way to merge these static/archived list of Subjects with the freshly-fetched list of Subjects for the other active projects, but this should be solvable.

### Dev Notes

Q: Anything to be cautious about? ⚠️ 
A: (1) are there any reasons that Subjects shouldn't saved on a publicly available repo? (e.g. sensitive data, copyright) (2) is this scalable with additional projects? i.e. how huge would the list of "archived" Subject CSVs/ZIPs be on this repo?

Q: Why does efficiency matter right now?
A: I'm planning to add the [Community Catalog's](https://github.com/zooniverse/community-catalog) (CC) projects to this database service, which would only increase SSSApi's boot up time. As a result, improving the time spent booting up (for development) and rebuilding (when CC's projects inevitably add new Subjects) the SSSApi may be a good idea.

### Status

Exploratory phase. Please add comments.

**Comment by eatyourgreens** [comment]
As we’ve discussed on Slack, Community Catalogue subjects shouldn’t be read from the API anyway. 

Instead, export the subjects from the project builder, then commit the subjects export to this repo. You’ll need to write a script to parse the metadata out of the export, but that should be doable.

**Reply by shaunanoordin** [reply]
For Community Catalogue, I would actually like to read the data from the API. Reading from a repo-stored ZIP/CSV (manual process) is something I've got as a Plan B, but I'm working under the understanding that the CC / Communities & Crowds project(s?) will require updates often enough that a semi-automated process (i.e. something Sam or me or anyone can trigger with a simple "lita rebuild") would be an acceptable tradeoff between build time and manual work.

But yeah, the scripting for either "extract metadata from Subject Export CSVs" or "extract metadata from Subjects straight out of Panoptes" would be very doable, as for CC/C&C we would knows these metadata fields in advance.

**Reply by eatyourgreens** [reply]
Subjects are linked to projects via workflows, but maybe you can query the subjects endpoint by project ID. 

However, with a page size of 20, that’s still 5,000 requests for a project with 100k subjects.

**Reply by eatyourgreens** [reply]
The Python client, as far as I can tell, fetches the subjects one-by-one from the API. It would be useful to find out how many API requests it’s making.

**Comment by eatyourgreens** [comment]
#26 moved the data-fetching step from the build into its own Docker image. That’s a step in the right direction. The database build is separate from the process of building the CSVs, and the CSVs could be persisted across builds.

**Comment by eatyourgreens** [comment]
I’d also suggest using Node to build the data files. I think async paged requests with the JS client will be faster than Python.

**Reply by eatyourgreens** [reply]
Keep an eye on this branch for (hopefully) much faster data-fetching.
https://github.com/zooniverse/subject-set-search-api/tree/migrate-to-node

**Reply by eatyourgreens** [reply]
Now that the image builds with Node, not Python, build times are less than a minute for the CSV files. See #30.

**Reply by eatyourgreens** [reply]
I set the subject page size to 50, but I think Panoptes allows page sizes up to 100 items. 

It might be worth tuning the query page size to see what effect that has on performance.

**Reply by eatyourgreens** [reply]
#35 increased the page size to 100. Build times haven’t changed much.
